File: engineering/concurrency/12-context-cancellation/skeleton/python/main.py
# 并发模型：threading + Event
import threading
import time

class CancelContext:
    def __init__(self):
        # TODO: 初始化取消事件和线程计数器/锁
        pass

    def cancel(self):
        # TODO: 设置取消事件，等待所有子线程结束
        pass

    def go(self, f):
        # TODO: 增加计数器，启动线程，传入取消事件；线程退出时减少计数器
        pass

def worker(event, wid):
    # 示例 worker 函数
    try:
        while not event.is_set():
            time.sleep(0.05)
        print(f"worker {wid} cancelled")
    finally:
        pass  # TODO: 确保退出时通知 context

def main():
    ctx = CancelContext()
    active = [0]
    mu = threading.Lock()

    for i in range(5):
        def make_worker(wid):
            def w():
                with mu:
                    active[0] += 1
                try:
                    while not ctx._event.is_set():  # 直接访问 event 仅为示例
                        time.sleep(0.05)
                    print(f"worker {wid} cancelled")
                finally:
                    with mu:
                        active[0] -= 1
            return w
        ctx.go(make_worker(i))

    time.sleep(0.1)
    ctx.cancel()

    if active[0] == 0:
        print("PASS: all workers exited after cancellation")
    else:
        print(f"FAIL: {active[0]} workers still active after cancellation")

if __name__ == "__main__":
    main()

File: engineering/concurrency/12-context-cancellation/skeleton/go/main.go
package main

import (
	"fmt"
	"sync"
	"time"
)

// CancelContext 简化版可取消上下文。
// TODO: 添加取消 channel 和 WaitGroup（或 done channel）。
type CancelContext struct {
	// TODO
}

// NewCancelContext 创建上下文。
func NewCancelContext() *CancelContext {
	// TODO: 初始化
	return &CancelContext{}
}

// Cancel 触发取消，并等待所有子任务退出。
func (ctx *CancelContext) Cancel() {
	// TODO: 关闭取消信号，等待所有子任务结束
}

// Go 启动一个受上下文管理的 goroutine。
func (ctx *CancelContext) Go(f func(done <-chan struct{})) {
	// TODO: 增加 WaitGroup，启动 goroutine 传入 done channel
}

func main() {
	ctx := NewCancelContext()
	active := 0
	var mu sync.Mutex

	for i := 0; i < 5; i++ {
		id := i
		ctx.Go(func(done <-chan struct{}) {
			mu.Lock()
			active++
			mu.Unlock()
			defer func() {
				mu.Lock()
				active--
				mu.Unlock()
			}()

			ticker := time.NewTicker(50 * time.Millisecond)
			defer ticker.Stop()
			for {
				select {
				case <-done:
					fmt.Printf("worker %d cancelled\n", id)
					return
				case <-ticker.C:
					// working
				}
			}
		})
	}

	// 给 worker 一点时间启动
	time.Sleep(100 * time.Millisecond)
	ctx.Cancel()

	mu.Lock()
	remaining := active
	mu.Unlock()
	if remaining == 0 {
		fmt.Println("PASS: all workers exited after cancellation")
	} else {
		fmt.Printf("FAIL: %d workers still active after cancellation\n", remaining)
	}
}
